civis/ml/_model.py

```diff
@@ -0,0 +1,443 @@
+        if client is None:
+            client = APIClient(resources='all')
+        super().__init__(client.scripts.get_containers_runs,
+                         [job_id, run_id],
+                         polling_interval=polling_interval,
+                         client=client,
+                         poll_on_creation=poll_on_creation)
+        if train_job_id and train_run_id:
+            self.is_training = False
+            self.train_job_id = train_job_id
+            self.train_run_id = train_run_id
+        else:
+            self.is_training = True
+            self.train_job_id = self.job_id
+            self.train_run_id = self.run_id
+        self._metadata, self._val_metadata = None, None
+        self._train_data, self._train_data_fname = None, None
+        self._train_metadata = None
+        self._table, self._estimator = None, None
+        self.add_done_callback(self._set_model_exception)
+
+    @property
+    def job_id(self):
+        return self.poller_args[0]
+
+    @property
+    def run_id(self):
+        return self.poller_args[1]
+
+    @staticmethod
+    def _set_model_exception(fut):
+        """Callback: On job completion, check the metadata.
+        If it indicates an exception, replace the generic
+        ``CivisJobFailure`` by a more informative ``ModelError``.
+        """
+        try:
+            meta = fut.metadata
+            if fut.is_training and meta['run']['status'] == 'succeeded':
+                # if training job and job succeeded, check validation job
+                meta = fut.validation_metadata
+            if (meta['run']['status'] == 'exception' and
+                    not isinstance(fut._exception, ModelError)):
+                # `set_exception` invokes callbacks, so make sure
+                # we haven't already set a `ModelError` to avoid
+                # infinite recursion.
+                try:
+                    # This will fail if the user doesn't have joblib installed
+                    est = fut.estimator
+                except Exception:  # NOQA
+                    est = None
+                fut.set_exception(
+                      ModelError('Model run failed with stack trace:\n'
+                                 '{}'.format(meta['run']['stack_trace']),
+                                 est, meta))
+        except (FileNotFoundError, CivisJobFailure):
+            # If there's no metadata file
+            # (we get FileNotFound or CivisJobFailure),
+            # then there's no improvements to make on
+            # any existing exception.
+            pass
+        except KeyError:
+            # KeyErrors always represent a bug in the modeling code,
+            # but showing the resulting KeyError can be confusing and
+            # mask the real error.
+            warnings.warn("Received malformed metadata from Civis Platform. "
+                          "Something went wrong with execution. "
+                          "Please report this error.")
```

Comment:
What's going on here?  Can we directly tell if this issue has occurred? I would prefer not to ask users to be reporting errors if we can avoid it.